geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/DescriptiveStats.scala

```diff
@@ -0,0 +1,358 @@
+  private[stats] val _mean: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _m2n: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _m3n: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _m4n: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _c2: SimpleMatrix = new SimpleMatrix(size, size)
+
+  clear()
+
+  override def clear(): Unit = {
+    _count = 0
+    _min.set(java.lang.Double.MAX_VALUE)
+    _max.set(0d - java.lang.Double.MAX_VALUE)
+    _sum.set(0d)
+    _mean.set(1d)
+    _m2n.set(0d)
+    _m3n.set(0d)
+    _m4n.set(0d)
+    _c2.set(0d)
+  }
+
+  def copyFrom(that: DescriptiveStats): Unit = {
+    _count = that._count
+    _min.set(that._min)
+    _max.set(that._max)
+    _sum.set(that._sum)
+    _mean.set(that._mean)
+    _m2n.set(that._m2n)
+    _m3n.set(that._m3n)
+    _m4n.set(that._m4n)
+    _c2.set(that._c2)
+  }
+
+  def count: Long = _count
+
+  def min: Array[Double] = (if (isEmpty) _max else _min).getMatrix.data.clone()
+  def max: Array[Double] = (if (isEmpty) _min else _max).getMatrix.data.clone()
+  def bounds: Array[(Double, Double)] = min.zip(max)
+
+  def sum: Array[Double] = _sum.getMatrix.data.clone()
+
+  def mean: Array[Double] = requireCount(1) { _mean }
+
+  def m2: Array[Double] = requireCount(1) { _m2n / _count }
+  def m3: Array[Double] = requireCount(1) { _m3n / _count }
+  def m4: Array[Double] = requireCount(1) { _m4n / _count }
+
+  def pvar: Array[Double] = requireCount(1) { _m2n / _count }
+  def psdev: Array[Double] = requireCount(1) { (_m2n / _count) ** 0.5 }
+  def pskew: Array[Double] = requireCount(1) { Math.sqrt(_count) * _m3n / (_m2n ** 1.5)  }
+  def pkurt: Array[Double] = requireCount(1) { _m4n * _count / (_m2n ** 2.0) }
+  def pekurt: Array[Double] = pkurt.map(_ -3.0 )
+
+  def svar: Array[Double] = requireCount(2) { _m2n / (_count - 1) }
+  def ssdev: Array[Double] = requireCount(2) { (_m2n / (_count - 1)) ** 0.5 }
+  def sskew: Array[Double] = requireCount(3) { _m3n * (_count * Math.sqrt(_count - 1) / (_count - 2)) / (_m2n ** 1.5) }
+  def skurt: Array[Double] = requireCount(4) {
+    _m4n * (count * (_count + 1) * (_count - 1) / (_count - 2) / (_count - 3)) / (_m2n ** 2.0)
+  }
+  def sekurt: Array[Double] = skurt.map(_ -3.0 )
+  
+  def c2: Array[Double] = requireCount(2, size_squared) { _c2 }
+
+  def pcov: Array[Double] = requireCount(2, size_squared) { _c2 / _count}
+  def pcor: Array[Double] = requireCount(2, size_squared) {
+    val mn2_sqrt = _m2n ** 0.5; (_c2 / (mn2_sqrt |*| mn2_sqrt.T)).diag(1.0) }
+
+  def scov: Array[Double] = requireCount(2, size_squared) { _c2 / (_count -1 ) }
+  def scor: Array[Double] = pcor /* population and sample calculations are equal w/ df term cancellation */
+
+  private def requireCount(count: Int, length: Int = size)(op: => SimpleMatrix):Array[Double] =
+    if (_count < count) Array.fill(length)(Double.NaN) else op.getMatrix.data.clone()
```

Comment:
so is this effectively saying if we don't have enough data, just return an Array of NaN, otherwise calculate the op?